src/Mod/Part/App/PartFeature.cpp

```diff
@@ -757,3 +757,235 @@ bool Part::checkIntersection(const TopoDS_Shape& first, const TopoDS_Shape& seco
     }
 
 }
+
+/**
+ * Override getElementName to support the Export type.  Other calls are passed to the original
+ * method
+ * @param name The name to search for, or if non existent, name of current Feature is returned
+ * @param type An element type name.
+ * @return The element name located, of
+ */
+std::pair<std::string, std::string> Feature::getElementName(const char* name,
+                                                            ElementNameType type) const
+{
+    if (type != ElementNameType::Export) {
+        return App::GeoFeature::getElementName(name, type);
+    }
+
+    // This function is overridden to provide higher level shape topo names that
+    // are generated on demand, e.g. Wire, Shell, Solid, etc.
+
+    auto prop = Base::freecad_dynamic_cast<PropertyPartShape>(getPropertyOfGeometry());
+    if (!prop) {
+        return App::GeoFeature::getElementName(name, type);
+    }
+
+    TopoShape shape = prop->getShape();
+    Data::MappedElement mapped = shape.getElementName(name);
+    auto res = shape.shapeTypeAndIndex(mapped.index);
+    static const int MinLowerTopoNames = 3;
+    static const int MaxLowerTopoNames = 10;
+    if (res.second && !mapped.name) {
+        // Here means valid index name, but no mapped name, check to see if
+        // we shall generate the high level topo name.
+        //
+        // The general idea of the algorithm is to find the minimum number of
+        // lower elements that can identify the given higher element, and
+        // combine their names to generate the name for the higher element.
+        //
+        // In theory, all it takes to find one lower element that only appear
+        // in the given higher element. To make the algorithm more robust
+        // against model changes, we shall take minimum MinLowerTopoNames lower
+        // elements.
+        //
+        // On the other hand, it may be possible to take too many elements for
+        // disambiguation. We shall limit to maximum MaxLowerTopoNames. If the
+        // chosen elements are not enough to disambiguate the higher element,
+        // we'll include an index for disambiguation.

```

Comment:
Can you add rewriting this comment to you TODO list? Not for this PR, but as future cleanup. It's pretty hard to parse.

Reply:
Done.